pytensor/link/numba/dispatch/elemwise_codegen.py

```diff
@@ -0,0 +1,240 @@
+from typing import Any, List, Optional, Tuple
```

Comment:
I can guess the answer to this question... but do we need to go this level? Can't we add some shape asserts in a vanilla numba function as a way to get the speedup without us having to do all this?

Reply:
Some shape asserts won't help I'm afraid. :-)
I think it is pretty clear that we want support for reductions during elemwise, and we also want to support multiple outputs. This means that the build-in numba vectorize won't do it.
So as far as I can see we have two options to get those:
- Build numba functions using string processing
- Use the llvm interface

I think working with strings is actually harder and more error-prone than working with the llvm code directly. (have a look at the Reduction functions in this module. I think that's even more complicated). Also, if we think about what happens if we do this, this would look like:

- We start with a properly typed graph
- We (conditionally) generate strings of numba code, that have no type info
- The numba code get's evaluated and transformed to bytecode
- The bytecode in analysed by numba
- Numba tries to figure out types for all the variables (those types that we discarded earlier, hopefully)
- Numba uses code just like in the PR to build an llvm module, based on the bytecode

I don't see why we would want those in-between steps to happen? `pytensor` is a compiler, I don't think we can reasonably expect to get around doing what compilers do: generate code. ;-)

Reply:
The long term problem I see here is maintenance. This is the backbone of the library and I don't know how much time it would take me (or another dev) to sort out a bug in the future. I don't know how volatile these functions are or how well documented.

Reply:
Yeah, the move from "we build a stats library" to "let's build a compiler" is a bit sudden :-)